### PR TITLE
GH-34752: [C++] Add support for LoongArch

### LINKED ISSUE #34752: [C++] Failed building on LoongArch platform
### Describe the bug, including details regarding any error messages, version, and platform.

```
loongson@loongson-pc:~/tmp/arrow/cpp/build$ arch
loongarch64
loongson@loongson-pc:~/tmp/arrow/cpp/build$ cmake ..
-- Building using CMake version: 3.13.4
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Arrow version: 12.0.0 (full: '12.0.0-SNAPSHOT')
-- Arrow SO version: 1200 (full: 1200.0.0)
-- clang-tidy 14 not found
-- clang-format 14 not found
-- Could NOT find ClangTools (missing: CLANG_FORMAT_BIN CLANG_TIDY_BIN) 
-- infer not found
-- Found Python3: /usr/local/bin/python3.8 (found version "3.8.2") found components:  Interpreter 
-- Found cpplint executable at /home/loongson/tmp/aa/arrow/cpp/build-support/cpplint.py
-- System processor: loongarch64
CMake Error at cmake_modules/SetupCxxFlags.cmake:40 (message):
  Unknown system processor
Call Stack (most recent call first):
  CMakeLists.txt:416 (include)


-- Configuring incomplete, errors occurred!
See also "/home/loongson/tmp/aa/arrow/cpp/build/CMakeFiles/CMakeOutput.log".

loongson@loongson-pc:~/tmp/arrow/cpp/build$ make
In file included from /home/loongson/tmp/aa/arrow/cpp/src/arrow/vendored/double-conversion/double-conversion.h:31,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/util/double_conversion.h:20,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/util/formatting.h:35,
                 from /home/loongson/tmp/aa/arrow/cpp/src/arrow/pretty_print.cc:41:
/home/loongson/tmp/aa/arrow/cpp/src/arrow/vendored/double-conversion/utils.h:118:2: error: #error Target architecture was not detected as supported by Double-Conversion.
 #error Target architecture was not detected as supported by Double-Conversion.
  ^~~~~
make[2]: *** [src/arrow/CMakeFiles/arrow_objlib.dir/build.make:466：src/arrow/CMakeFiles/arrow_objlib.dir/pretty_print.cc.o] 错误 1
make[1]: *** [CMakeFiles/Makefile2:684：src/arrow/CMakeFiles/arrow_objlib.dir/all] 错误 2
make: *** [Makefile:141：all] 错误 2

```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by liujiacheng777**
* Closes: #34752

### CODE REVIEW

**Comment by kou** [comment]
Could you open an issue for this pull request? This is not a minor change.

See also:

> If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose
> 
> Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.
> 
> Then could you also rename the pull request title in the following format?
> 
> ```
> GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}
> ```
>
> ...
>
> See also:
> 
>     * [Other pull requests](https://github.com/apache/arrow/pulls/)
> 
>     * [Contribution Guidelines - How to contribute patches](https://arrow.apache.org/docs/developers/contributing.html#how-to-contribute-patches)

**Comment by mapleFU** [comment]
By the way, do we have (or need have) CI for all the architectures? @kou

**Comment by kou** [comment]
It's better but we don't require it. Because it's difficult to have CI for all supported platforms. We can use QEMU on x86_64 but the approach has a performance problem.
For example, we have some s390x CI jobs on Travis CI but we'll abandon them because ASF stopped Travis CI support.

@liujiacheng777 Could you run C++ unit tests on your environment because we can't do it in CI?
https://arrow.apache.org/docs/dev/developers/cpp/development.html#running-unit-tests
(You also need to read https://arrow.apache.org/docs/dev/developers/cpp/index.html to run unit tests.)

File: cpp/src/arrow/vendored/double-conversion/utils.h [hunk 1]
```diff
@@ -102,6 +102,7 @@ int main(int argc, char** argv) {
     defined(__AARCH64EL__) || defined(__aarch64__) || defined(__AARCH64EB__) || \
     defined(__riscv) || \
     defined(__or1k__) || defined(__arc__) || \
+    defined(__loongarch__) || \
```
Comment (kou) [inline cpp/src/arrow/vendored/double-conversion/utils.h:105]: Could you open a pull request to upstream https://github.com/google/double-conversion ?
We'll update our bundled version after this change is merged in upstream.

Comment (loongson-zn) [reply cpp/src/arrow/vendored/double-conversion/utils.h:105]: https://github.com/google/double-conversion/pull/162/

Comment (kou) [reply cpp/src/arrow/vendored/double-conversion/utils.h:105]: Thanks.

Then could you open a new separated issue to update vendored double-conversion and open a separated pull request for it?
For example, #14817 is a pull request to update vendored fast_float.

**Comment by liujiacheng777** [comment]
https://github.com/apache/arrow/pull/34793

**Comment by liujiacheng777** [comment]
https://github.com/apache/arrow/pull/34793  
I have individually modified as required，Please review

**Comment by amol-** [comment]
This seems to be blocked by https://github.com/apache/arrow/pull/34793 as far as I can understand. But the other one is closed, if it's just updating the version of a vendored lib and the tests are passing, should we merge it and unblock this one?

**Comment by kou** [comment]
This is blocked by #34919.
I want to resolve https://github.com/apache/arrow/pull/34919#issuecomment-1501420706 before we merge #34919.

**Comment by loongson-zn** [comment]
I see that https://github.com/apache/arrow/pull/34919#issue-1656545063  already been merged !

**Comment by kou** [comment]
Could you rebase on main?

**Review by loongson-zn — APPROVED** [review]

**Review by kou — APPROVED** [review]
+1

**Pull request merged**